overlord/devicestate/users.go

```diff
@@ -0,0 +1,359 @@
+
+	// check that the signer of the assertion is one of the accepted ones
+	sysUserAuths := modelAs.SystemUserAuthority()
+	if len(sysUserAuths) > 0 && !strutil.ListContains(sysUserAuths, su.AuthorityID()) {
+		return "", nil, fmt.Errorf(errorPrefix+"%q not in accepted authorities %q", email, su.AuthorityID(), sysUserAuths)
+	}
+	if len(su.Series()) > 0 && !strutil.ListContains(su.Series(), series) {
+		return "", nil, fmt.Errorf(errorPrefix+"%q not in series %q", email, series, su.Series())
+	}
+	if len(su.Models()) > 0 && !strutil.ListContains(su.Models(), model) {
+		return "", nil, fmt.Errorf(errorPrefix+"%q not in models %q", model, su.Models())
+	}
+	if len(su.Serials()) > 0 {
+		if serialAs == nil {
+			return "", nil, fmt.Errorf(errorPrefix + "bound to serial assertion but device not yet registered")
+		}
+		serial := serialAs.Serial()
+		if !strutil.ListContains(su.Serials(), serial) {
+			return "", nil, fmt.Errorf(errorPrefix+"%q not in serials %q", serial, su.Serials())
+		}
+	}
+
+	if !su.ValidAt(time.Now()) {
+		return "", nil, fmt.Errorf(errorPrefix + "assertion not valid anymore")
+	}
+
+	gecos := fmt.Sprintf("%s,%s", email, su.Name())
+	opts := &osutil.AddUserOptions{
+		SSHKeys:             su.SSHKeys(),
+		Gecos:               gecos,
+		Password:            su.Password(),
+		ForcePasswordChange: su.ForcePasswordChange(),
+	}
+	return su.Username(), opts, nil
+}
+
+func (u *createUserOpts) setupLocalUser(state *state.State, username, email string) error {
+	user, err := userLookup(username)
+	if err != nil {
+		return fmt.Errorf("cannot lookup user %q: %s", username, err)
+	}
+	uid, gid, err := osutil.UidGid(user)
+	if err != nil {
+		return err
+	}
+	authDataFn := filepath.Join(user.HomeDir, ".snap", "auth.json")
+	if err := osutil.MkdirAllChown(filepath.Dir(authDataFn), 0700, uid, gid); err != nil {
+		return err
+	}
+
+	// setup new user, local-only
+	authUser, err := auth.NewUser(state, auth.NewUserData{
+		Username:   username,
+		Email:      email,
+		Macaroon:   "",
+		Discharges: nil,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot persist authentication details: %v", err)
+	}
+	// store macaroon auth, user's ID, email and username in auth.json in
+	// the new users home dir
+	outStr, err := json.Marshal(struct {
+		ID       int    `json:"id"`
+		Username string `json:"username"`
+		Email    string `json:"email"`
+		Macaroon string `json:"macaroon"`
+	}{
+		ID:       authUser.ID,
+		Username: authUser.Username,
+		Email:    authUser.Email,
+		Macaroon: authUser.Macaroon,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot marshal auth data: %s", err)
+	}
+	if err := osutil.AtomicWriteFileChown(authDataFn, []byte(outStr), 0600, 0, uid, gid); err != nil {
+		return fmt.Errorf("cannot write auth file %q: %s", authDataFn, err)
+	}
+
+	return nil
+}
+
+func (u *createUserOpts) addUser(state *state.State, username string, email string, opts *osutil.AddUserOptions) (CreatedUser, error) {
```

Comment:
Likewise, if the function above wasn't a method, this also wouldn't have to be.

Reply:
Fixed!